packages/server/src/accounts-server.ts

```diff
@@ -90,60 +90,59 @@ export class AccountsServer {
     params,
     infos: ConnectionInformations
   ): Promise<LoginResult> {
-    if (!this.services[serviceName]) {
-      throw new Error(`No service with the name ${serviceName} was registered.`);
-    }
-    const user: User = await this.services[serviceName].authenticate(params);
-    if (!user) {
-      throw new Error(`Service ${serviceName} was not able to authenticate user`);
-    }
-
-    await this.hooks.emitSerial(ServerHooks.Login, {
+    const hooksInfo: any = {
       // The service name, such as “password” or “twitter”.
       service: serviceName,
-      // The user object
-      user,
       // The connection informations <ConnectionInformations>
       connection: infos,
-    });
-    return this.loginWithUser(user, infos);
+    };
+    try {
+      if (!this.services[serviceName]) {
+        throw new Error(`No service with the name ${serviceName} was registered.`);
+      }
+
+      const user: User = await this.services[serviceName].authenticate(params);
+      if (!user) {
+        throw new Error(`Service ${serviceName} was not able to authenticate user`);
+      }
+      hooksInfo.user = user;
+
+      // Let the user validate the login attempt
+      await this.hooks.emitSerial(ServerHooks.ValidateLogin, hooksInfo);
+      const loginResult = await this.loginWithUser(user, infos);
+      this.hooks.emit(ServerHooks.LoginSuccess, hooksInfo);
+      return loginResult;
+    } catch (err) {
+      this.hooks.emit(ServerHooks.LoginError, hooksInfo);
+      throw err;
+    }
   }
 
   /**
-   * @description Server use only. This method creates a session
-   *              without authenticating any user identity.
-   *              Any authentication should happen before calling this function.
+   * @description Server use only.
+   * This method creates a session without authenticating any user identity.
+   * Any authentication should happen before calling this function.
    * @param {User} userId - The user object.
    * @param {string} ip - User's ip.
    * @param {string} userAgent - User's client agent.
    * @returns {Promise<LoginResult>} - Session tokens and user object.
    */
   public async loginWithUser(user: User, infos: ConnectionInformations): Promise<LoginResult> {
```

Comment:
As this is a public method, some users that call loginWithUser directly (maybe even @dotansimha?) will not have the login and error hooks called. This is a breaking change, are we ready for minor release?

Reply:
Hum true, I was wondering is this function should trigger or not the hooks since it can only be called server side?

Reply:
Yeah I think it is currently one of the more flexible ways to create your own authentication methods outside of our auth-services mechanism. I don't mind simplifying this code, but it is breaking..

Reply:
So the problem is that we currently need to have more info inside the hooks can you think about a better approach?